DllImportGenerator/DllImportGenerator/DllImportGenerator.cs

```diff
@@ -1,164 +1,183 @@
+                    static (node, ct) => ShouldVisitNode(node),
+                    static (context, ct) =>
+                        new 
+                        {
+                            Syntax = (MethodDeclarationSyntax)context.Node,
+                            Symbol = (IMethodSymbol)context.SemanticModel.GetDeclaredSymbol(context.Node, ct)!
+                        })
+                .Where(
+                    static modelData => modelData.Symbol.IsStatic && modelData.Symbol.GetAttributes().Any(
+                        static attribute => attribute.AttributeClass?.ToDisplayString() == TypeNames.GeneratedDllImportAttribute)
+                );
+
+            var compilationAndTargetFramework = context.CompilationProvider
+                .Select(static (compilation, ct) =>
                 {
-                    sm = context.Compilation.GetSemanticModel(methodSyntax.SyntaxTree, ignoreAccessibility: true);
-                    syntaxToModel.Add(methodSyntax.SyntaxTree, sm);
-                }
-
-                // Process the method syntax and get its SymbolInfo.
-                var methodSymbolInfo = sm.GetDeclaredSymbol(methodSyntax, context.CancellationToken)!;
-
-                // Get any attributes of interest on the method
-                AttributeData? generatedDllImportAttr = null;
-                AttributeData? lcidConversionAttr = null;
-                AttributeData? suppressGCTransitionAttribute = null;
-                AttributeData? unmanagedCallConvAttribute = null;
-
-                foreach (var attr in methodSymbolInfo.GetAttributes())
+                    bool isSupported = IsSupportedTargetFramework(compilation, out Version targetFrameworkVersion);
+                    return (compilation, isSupported, targetFrameworkVersion);
+                });
+
+            context.RegisterSourceOutput(
+                compilationAndTargetFramework
+                    .Combine(methodsToGenerate.Collect()),
+                static (context, data) =>
                 {
-                    if (attr.AttributeClass is null)
+                    if (!data.Left.isSupported && data.Right.Any())
                     {
-                        continue;
+                        // We don't block source generation when the TFM is unsupported.
+                        // This allows a user to copy generated source and use it as a starting point
+                        // for manual marshalling if desired.
+                        context.ReportDiagnostic(
+                            Diagnostic.Create(
+                                GeneratorDiagnostics.TargetFrameworkNotSupported,
+                                Location.None,
+                                MinimumSupportedFrameworkVersion.ToString(2)));
                     }
-                    else if (attr.AttributeClass.ToDisplayString() == TypeNames.GeneratedDllImportAttribute)
+                });
+
+            var stubEnvironment = compilationAndTargetFramework
```

Comment:
This logic needs extensive comments below. What is `Left` and `Right`?

Reply:
Left and Right are the names provided for the values returned by the `Combine` method. The `Left` member is the data on the left side of the `Combine` method (the `this` parameter), and `Right` is on the right side of the `Combine` method (the parameter passed directly to the method).

Do you want me to add explicit comments about the types of `Left` and `Right`? I'm a little apprehensive about that since it would get out of date quickly if we change the pipeline at all.

Reply:
This is simply ignorance on my part. I've avoided Linq functions for years due to overhead compared to explicit logic - I am assuming `Combine` is some Linq function? I would appreciate a comment for each independent statement in this function. There are baked in assumptions here about how we efficiently work with Roslyn and I'd like to ensure we state those assumptions or at least describe the thought process for how we got here. This implementation, at least to me, feels like there is some opportunity for the IDE team to improve the contract. Will all source generators implementations need to think about this if they want to hit IDE performance metrics? If so, is the methodology documented or captured in an official example?

Reply:
The `Select`, `WithComparer`, `Combine`, and `Where` methods here aren't actually LINQ methods. They're provided by Roslyn on the `IncrementalValue(s)Provider` types and are named like the LINQ methods for consistency with the BCL. They're actually the only way to do any sort of manipulation of the provider types to define an incremental generator pipeline.


See the source in Roslyn: 
https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/IncrementalValueProvider.cs

https://github.com/dotnet/roslyn/blob/main/src/Compilers/Core/Portable/SourceGeneration/Nodes/ValueSourceExtensions.cs

Reply:
>They're provided by Roslyn on the IncrementalValue(s)Provider types and are named like the LINQ methods for consistency with the BCL.

Ah. This is one of those things where it would have been obvious if I loaded it in VS or another IDE. Thanks for letting me know. It is annoying that C# requires an IDE to actually do a good review :-(